// @nrweb/astro-calc/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RETROGRADE_PLANETS = exports.PLANET_ORBITS = exports.EARTH_YEAR = exports.EARTH_ORBIT_RADIUS = exports.PLANET_SPEED_LOWER_BOUND = exports.PLANETS = exports.BOUNDS = exports.DECANS = exports.PLANET_SECTS = exports.PLANET_DIGNITIES = exports.CHALDEAN_ORDER = exports.DAY_RULERS = exports.ZODIAC_SIGNS = exports.DEFAULT_LONGITUDE = exports.DEFAULT_LATITUDE = exports.SE_PLANETS = void 0;
const astronomy_engine_1 = require("astronomy-engine");
// Swiss Ephemeris planet constants
exports.SE_PLANETS = {
    [astronomy_engine_1.Body.Sun]: 0, // SE_SUN
    [astronomy_engine_1.Body.Moon]: 1, // SE_MOON
    [astronomy_engine_1.Body.Mercury]: 2, // SE_MERCURY
    [astronomy_engine_1.Body.Venus]: 3, // SE_VENUS
    [astronomy_engine_1.Body.Mars]: 4, // SE_MARS
    [astronomy_engine_1.Body.Jupiter]: 5, // SE_JUPITER
    [astronomy_engine_1.Body.Saturn]: 6, // SE_SATURN
    [astronomy_engine_1.Body.Uranus]: 7, // SE_URANUS
    [astronomy_engine_1.Body.Neptune]: 8, // SE_NEPTUNE
    [astronomy_engine_1.Body.Pluto]: 9, // SE_PLUTO
};
// Default coordinates for Belfast, UK
exports.DEFAULT_LATITUDE = 54.5973;
exports.DEFAULT_LONGITUDE = -5.9301;
// Zodiac signs in order
exports.ZODIAC_SIGNS = [
    "Aries",
    "Taurus",
    "Gemini",
    "Cancer",
    "Leo",
    "Virgo",
    "Libra",
    "Scorpio",
    "Sagittarius",
    "Capricorn",
    "Aquarius",
    "Pisces",
];
exports.DAY_RULERS = [
    astronomy_engine_1.Body.Sun,
    astronomy_engine_1.Body.Moon,
    astronomy_engine_1.Body.Mars,
    astronomy_engine_1.Body.Mercury,
    astronomy_engine_1.Body.Jupiter,
    astronomy_engine_1.Body.Venus,
    astronomy_engine_1.Body.Saturn,
];
exports.CHALDEAN_ORDER = [
    astronomy_engine_1.Body.Sun,
    astronomy_engine_1.Body.Venus,
    astronomy_engine_1.Body.Mercury,
    astronomy_engine_1.Body.Moon,
    astronomy_engine_1.Body.Saturn,
    astronomy_engine_1.Body.Jupiter,
    astronomy_engine_1.Body.Mars,
];
// Traditional astrological dignities
exports.PLANET_DIGNITIES = {
    [astronomy_engine_1.Body.Sun]: {
        rulership: "Leo",
        exaltation: "Aries",
        detriment: "Aquarius",
        fall: "Libra",
    },
    [astronomy_engine_1.Body.Moon]: {
        rulership: "Cancer",
        exaltation: "Taurus",
        detriment: "Capricorn",
        fall: "Scorpio",
    },
    [astronomy_engine_1.Body.Mercury]: {
        rulership: ["Gemini", "Virgo"],
        exaltation: "Virgo",
        detriment: ["Sagittarius", "Pisces"],
        fall: "Pisces",
    },
    [astronomy_engine_1.Body.Venus]: {
        rulership: ["Taurus", "Libra"],
        exaltation: "Pisces",
        detriment: ["Aries", "Scorpio"],
        fall: "Virgo",
    },
    [astronomy_engine_1.Body.Mars]: {
        rulership: ["Aries", "Scorpio"],
        exaltation: "Capricorn",
        detriment: ["Taurus", "Libra"],
        fall: "Cancer",
    },
    [astronomy_engine_1.Body.Jupiter]: {
        rulership: ["Sagittarius", "Pisces"],
        exaltation: "Cancer",
        detriment: ["Gemini", "Virgo"],
        fall: "Capricorn",
    },
    [astronomy_engine_1.Body.Saturn]: {
        rulership: ["Capricorn", "Aquarius"],
        exaltation: "Libra",
        detriment: ["Cancer", "Leo"],
        fall: "Aries",
    },
};
// Planet sect information
exports.PLANET_SECTS = {
    [astronomy_engine_1.Body.Sun]: "Diurnal",
    [astronomy_engine_1.Body.Moon]: "Nocturnal",
    [astronomy_engine_1.Body.Mercury]: "Neutral",
    [astronomy_engine_1.Body.Venus]: "Nocturnal",
    [astronomy_engine_1.Body.Mars]: "Nocturnal",
    [astronomy_engine_1.Body.Jupiter]: "Diurnal",
    [astronomy_engine_1.Body.Saturn]: "Diurnal",
    [astronomy_engine_1.Body.Uranus]: "Neutral",
    [astronomy_engine_1.Body.Neptune]: "Neutral",
    [astronomy_engine_1.Body.Pluto]: "Neutral",
};
exports.DECANS = {
    Aries: ["Mars", "Sun", "Venus"],
    Taurus: ["Mercury", "Moon", "Saturn"],
    Gemini: ["Jupiter", "Mars", "Sun"],
    Cancer: ["Venus", "Mercury", "Moon"],
    Leo: ["Saturn", "Jupiter", "Mars"],
    Virgo: ["Sun", "Venus", "Mercury"],
    Libra: ["Moon", "Saturn", "Jupiter"],
    Scorpio: ["Mars", "Sun", "Venus"],
    Sagittarius: ["Mercury", "Moon", "Saturn"],
    Capricorn: ["Jupiter", "Mars", "Sun"],
    Aquarius: ["Venus", "Mercury", "Moon"],
    Pisces: ["Saturn", "Jupiter", "Mars"],
};
exports.BOUNDS = {
    Aries: [
        { start: 0, end: 6, planet: astronomy_engine_1.Body.Jupiter },
        { start: 7, end: 12, planet: astronomy_engine_1.Body.Venus },
        { start: 13, end: 20, planet: astronomy_engine_1.Body.Mercury },
        { start: 21, end: 25, planet: astronomy_engine_1.Body.Mars },
        { start: 26, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Taurus: [
        { start: 0, end: 8, planet: astronomy_engine_1.Body.Venus },
        { start: 9, end: 14, planet: astronomy_engine_1.Body.Mercury },
        { start: 15, end: 22, planet: astronomy_engine_1.Body.Jupiter },
        { start: 23, end: 27, planet: astronomy_engine_1.Body.Saturn },
        { start: 28, end: 30, planet: astronomy_engine_1.Body.Mars },
    ],
    Gemini: [
        { start: 0, end: 6, planet: astronomy_engine_1.Body.Mercury },
        { start: 7, end: 12, planet: astronomy_engine_1.Body.Jupiter },
        { start: 13, end: 17, planet: astronomy_engine_1.Body.Venus },
        { start: 18, end: 24, planet: astronomy_engine_1.Body.Mars },
        { start: 25, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Cancer: [
        { start: 0, end: 7, planet: astronomy_engine_1.Body.Mars },
        { start: 7, end: 13, planet: astronomy_engine_1.Body.Venus },
        { start: 13, end: 19, planet: astronomy_engine_1.Body.Mercury },
        { start: 20, end: 26, planet: astronomy_engine_1.Body.Mars },
        { start: 27, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Leo: [
        { start: 0, end: 6, planet: astronomy_engine_1.Body.Jupiter },
        { start: 7, end: 11, planet: astronomy_engine_1.Body.Venus },
        { start: 12, end: 18, planet: astronomy_engine_1.Body.Saturn },
        { start: 19, end: 24, planet: astronomy_engine_1.Body.Mercury },
        { start: 25, end: 30, planet: astronomy_engine_1.Body.Mars },
    ],
    Virgo: [
        { start: 0, end: 7, planet: astronomy_engine_1.Body.Mercury },
        { start: 8, end: 17, planet: astronomy_engine_1.Body.Venus },
        { start: 18, end: 21, planet: astronomy_engine_1.Body.Jupiter },
        { start: 22, end: 28, planet: astronomy_engine_1.Body.Mars },
        { start: 29, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Libra: [
        { start: 0, end: 6, planet: astronomy_engine_1.Body.Saturn },
        { start: 7, end: 14, planet: astronomy_engine_1.Body.Mercury },
        { start: 15, end: 21, planet: astronomy_engine_1.Body.Jupiter },
        { start: 22, end: 28, planet: astronomy_engine_1.Body.Venus },
        { start: 26, end: 30, planet: astronomy_engine_1.Body.Mars },
    ],
    Scorpio: [
        { start: 0, end: 7, planet: astronomy_engine_1.Body.Mars },
        { start: 8, end: 11, planet: astronomy_engine_1.Body.Venus },
        { start: 12, end: 19, planet: astronomy_engine_1.Body.Mercury },
        { start: 20, end: 24, planet: astronomy_engine_1.Body.Jupiter },
        { start: 25, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Sagittarius: [
        { start: 0, end: 12, planet: astronomy_engine_1.Body.Jupiter },
        { start: 13, end: 17, planet: astronomy_engine_1.Body.Venus },
        { start: 18, end: 21, planet: astronomy_engine_1.Body.Mercury },
        { start: 22, end: 26, planet: astronomy_engine_1.Body.Saturn },
        { start: 26, end: 30, planet: astronomy_engine_1.Body.Mars },
    ],
    Capricorn: [
        { start: 0, end: 7, planet: astronomy_engine_1.Body.Mercury },
        { start: 8, end: 14, planet: astronomy_engine_1.Body.Jupiter },
        { start: 15, end: 22, planet: astronomy_engine_1.Body.Venus },
        { start: 23, end: 26, planet: astronomy_engine_1.Body.Saturn },
        { start: 26, end: 30, planet: astronomy_engine_1.Body.Mars },
    ],
    Aquarius: [
        { start: 0, end: 7, planet: astronomy_engine_1.Body.Mercury },
        { start: 8, end: 13, planet: astronomy_engine_1.Body.Venus },
        { start: 14, end: 20, planet: astronomy_engine_1.Body.Jupiter },
        { start: 21, end: 25, planet: astronomy_engine_1.Body.Mars },
        { start: 26, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
    Pisces: [
        { start: 0, end: 12, planet: astronomy_engine_1.Body.Venus },
        { start: 13, end: 16, planet: astronomy_engine_1.Body.Jupiter },
        { start: 17, end: 19, planet: astronomy_engine_1.Body.Mercury },
        { start: 20, end: 28, planet: astronomy_engine_1.Body.Mars },
        { start: 29, end: 30, planet: astronomy_engine_1.Body.Saturn },
    ],
};
exports.PLANETS = [
    astronomy_engine_1.Body.Sun,
    astronomy_engine_1.Body.Moon,
    astronomy_engine_1.Body.Mercury,
    astronomy_engine_1.Body.Venus,
    astronomy_engine_1.Body.Mars,
    astronomy_engine_1.Body.Jupiter,
    astronomy_engine_1.Body.Saturn,
    astronomy_engine_1.Body.Uranus,
    astronomy_engine_1.Body.Neptune,
    astronomy_engine_1.Body.Pluto,
];
// Lower bounds for determining slow motion (planets moving slower than this are considered "slow")
exports.PLANET_SPEED_LOWER_BOUND = {
    [astronomy_engine_1.Body.Moon]: 12.3,
    [astronomy_engine_1.Body.Mercury]: 0.5,
    [astronomy_engine_1.Body.Venus]: 0.5,
    [astronomy_engine_1.Body.Mars]: 0.3,
    [astronomy_engine_1.Body.Jupiter]: 0.06,
    [astronomy_engine_1.Body.Saturn]: 0.02,
};
// Constants for orbital calculations
exports.EARTH_ORBIT_RADIUS = 1.0; // AU
exports.EARTH_YEAR = 365.24; // days
// Orbital data for planets we track
exports.PLANET_ORBITS = {
    [astronomy_engine_1.Body.Mercury]: { radius: 0.387, year: 87.97 },
    [astronomy_engine_1.Body.Venus]: { radius: 0.723, year: 224.7 },
    [astronomy_engine_1.Body.Mars]: { radius: 1.524, year: 686.98 },
    [astronomy_engine_1.Body.Jupiter]: { radius: 5.203, year: 4332.59 },
    [astronomy_engine_1.Body.Saturn]: { radius: 9.537, year: 10759.22 },
    [astronomy_engine_1.Body.Uranus]: { radius: 19.191, year: 30688.5 },
    [astronomy_engine_1.Body.Neptune]: { radius: 30.069, year: 60182 },
    [astronomy_engine_1.Body.Pluto]: { radius: 39.482, year: 90520 },
    [astronomy_engine_1.Body.Sun]: { radius: 0, year: 0 },
    [astronomy_engine_1.Body.Moon]: { radius: 0, year: 0 },
};
exports.RETROGRADE_PLANETS = [
    astronomy_engine_1.Body.Mercury,
    astronomy_engine_1.Body.Venus,
    astronomy_engine_1.Body.Mars,
    astronomy_engine_1.Body.Jupiter,
    astronomy_engine_1.Body.Saturn,
    astronomy_engine_1.Body.Uranus,
    astronomy_engine_1.Body.Neptune,
    astronomy_engine_1.Body.Pluto,
];
